src/libsyntax/parse/parser/item.rs

```diff
@@ -122,12 +122,12 @@ impl<'a> Parser<'a> {
             if self.eat_keyword(kw::Fn) {
                 // EXTERN FUNCTION ITEM
                 let fn_span = self.prev_span;
-                let abi = opt_abi.unwrap_or(Abi::C);
-                let (ident, item_, extra_attrs) =
-                    self.parse_item_fn(Unsafety::Normal,
-                                       respan(fn_span, IsAsync::NotAsync),
-                                       respan(fn_span, Constness::NotConst),
-                                       abi)?;
+                let (ident, item_, extra_attrs) = self.parse_item_fn(FnHeader {
```

Comment:
It looked less weird without gathering the arguments into a single struct, IMO.
I'd rather construct `FnHeader` inside `parse_item_fn`.

Reply:
I opted to reformat things a bit so that it hopefully looks less weird to you. I hope that's OK. Since I added more parameters, passing all of the `FnHeader` bits as well will make that too much to fit nicely.

My overall goal here is to make free- and associated-function parsing be more in-line as well as to eventually fuse the entirety of `FnHeader` parsing into `parse_item_fn` with `fn parse_fn_front_matter` by using look-ahead to check if it's OK (but that's for a different PR...).